go/signedexchange/cmd/dump-signedexchange/main.go

```diff
@@ -1,53 +1,118 @@
-	flagCert      = flag.String("cert", "", "Certificate CBOR file. If specified, used instead of fetching from signature's cert-url")
+	flagCert          = flag.String("cert", "", "Certificate CBOR file. If specified, used instead of fetching from signature's cert-url")
+	flagHeaders       = flag.Bool("headers", true, "Print headers")
+	flagFilename      = flag.String("i", "", "Signed-exchange input file")
+	flagPayload       = flag.Bool("payload", true, "Print payload")
+	flagSignature     = flag.Bool("signature", false, "Prints only signature value")
+	flagURI           = flag.String("uri", "", "Signed-exchange uri")
+	flagVerify        = flag.Bool("verify", false, "Perform signature verification")
+	flagVersion       = flag.String("version", latestVersion, "Signed exchange version")
+	flagRequestHeader = flag.String("requestHeader", "", "Request header for signed exchange")
 )
 
+func getAcceptHeader(value string) (string, error) {
+	v, parsed := version.Parse(value)
+	if parsed {
+		return fmt.Sprintf("application/signed-exchange;v=%s", v[1:]), nil
+	}
+	return "", fmt.Errorf("Could not parse version string \"%s\"", value)
+}
+
+func getRequestHeader(value string) ([]string, error) {
+	h := strings.SplitN(value, ":", 2)
+	if len(h) == 2 {
+		return h, nil
+	}
+	return nil, errors.New("Could not parse header. Should be \"key:value\"")
+}
+
 func run() error {
-	in := os.Stdin
-	if *flagInput != "" {
-		var err error
-		in, err = os.Open(*flagInput)
+	fi, err := os.Stdin.Stat()
+	if err != nil {
+		return err
+	}
+	var e *signedexchange.Exchange
+	var in io.Reader = nil
+	if (fi.Mode() & os.ModeCharDevice) == 0 { // read sxg from pipe
+		in = os.Stdin
+	} else if *flagFilename != "" { // read sxg from filename
+		in, err = os.Open(*flagFilename)
+		if err != nil {
+			return err
+		}
+	} else if *flagURI != "" { // read sxg from network
+		client := http.DefaultClient
+		req, err := http.NewRequest("GET", *flagURI, nil)
+		if err != nil {
+			return err
+		}
+		acceptHeader, err := getAcceptHeader(*flagVersion)
 		if err != nil {
 			return err
 		}
-		defer in.Close()
+		req.Header.Add("Accept", acceptHeader)
+		if *flagRequestHeader != "" {
+			requestHeader, err := getRequestHeader(*flagRequestHeader)
+			if err != nil {
+				return err
+			}
+			req.Header.Add(requestHeader[0], requestHeader[1])
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		in = resp.Body
+		defer resp.Body.Close()
 	}
 
-	e, err := signedexchange.ReadExchange(in)
+	if in == nil {
+		flag.PrintDefaults()
+		return nil
+	}
+
+	e, err = signedexchange.ReadExchange(in)
 	if err != nil {
 		return err
 	}
-
 	if *flagSignature {
 		fmt.Println(e.SignatureHeaderValue)
-		return nil
-	}
+	} else {
+		if *flagHeaders {
+			e.PrettyPrintHeaders(os.Stdout)
+		}
 
-	if *flagVerify {
-		if err := verify(e); err != nil {
-			return err
+		if *flagPayload {
+			e.PrettyPrintPayload(os.Stdout)
```

Comment:
Currently this tool prints mice-decoded payload if `-verify` is given (because `verify()` decodes the payload), but after this patch payload is not decoded at this point.

I know this is confusing and we should fix it (#431), but can we keep the current behavior for now?

Reply:
I believe this behavior is unchanged